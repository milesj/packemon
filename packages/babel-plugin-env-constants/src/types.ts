/* eslint-disable no-underscore-dangle */

declare global {
	const __DEV__: boolean;
	const __PROD__: boolean;
	const __TEST__: boolean;
}

export type {};
