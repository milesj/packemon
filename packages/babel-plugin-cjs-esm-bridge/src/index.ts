import paths from 'path';
import { NodePath, PluginObj, types as t } from '@babel/core';

function isEsmFile(file: string) {
	return /\.(ts|tsx|mjs)$/.test(file);
}

export default function cjsEsmBridge(): PluginObj {
	return {
		visitor: {
			CallExpression: {
				enter(path: NodePath<t.CallExpression>, state) {
					const isEsm = isEsmFile(state.filename);

					// require() not allowed in esm files
					// https://nodejs.org/api/esm.html#esm_no_require_exports_or_module_exports
					if (isEsm && path.get('callee').isIdentifier({ name: 'require' })) {
						throw new Error(
							`Found a \`require()\` call in "${paths.basename(
								state.filename,
							)}", this is not allowed in ".mjs" files. Use dynamic \`import()\` instead.`,
						);
					}
				},
			},
		},
	};
}
