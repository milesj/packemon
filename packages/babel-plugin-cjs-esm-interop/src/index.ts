/* eslint-disable complexity */

import paths from 'path';
import { NodePath, PluginObj, PluginPass, types as t } from '@babel/core';
// @ts-expect-error Not typed
import { addDefault } from '@babel/helper-module-imports';

export interface CjsEsmInteropOptions {
	// The output format
	format?: 'cjs' | 'mjs';
}

// function isEsmFile(file: string) {
// 	return /\.(ts|tsx|mjs)$/.test(file);
// }

function isPathDirname(path: NodePath): boolean {
	// dirname(foo)
	if (path.isCallExpression() && path.get('callee').isIdentifier({ name: 'dirname' })) {
		return true;
	}

	// path.dirname(foo)
	return Boolean(
		path.isCallExpression() &&
			path.get('callee').isMemberExpression() &&
			(path.get('callee.object') as NodePath).isIdentifier() &&
			(path.get('callee.property') as NodePath).isIdentifier({ name: 'dirname' }),
	);
}

function isProcessEnv(path: NodePath): boolean {
	// process.env
	return Boolean(
		path.isMemberExpression() &&
			(path.get('object') as NodePath).isMemberExpression() &&
			(path.get('object.object') as NodePath).isIdentifier({ name: 'process' }) &&
			(path.get('object.property') as NodePath).isIdentifier({ name: 'env' }),
	);
}

function getFormat(state: PluginPass): 'cjs' | 'mjs' {
	return (state.opts as CjsEsmInteropOptions)?.format ?? 'mjs';
}

function wrapWithFileProto(path: NodePath, id: string) {
	if (
		path.parentPath?.isNewExpression() &&
		path.parentPath?.get('callee').isIdentifier({ name: 'URL' })
	) {
		return t.binaryExpression('+', t.stringLiteral('file://'), t.identifier(id));
	}

	return t.identifier(id);
}

export default function cjsEsmInterop(): PluginObj {
	return {
		visitor: {
			CallExpression(path: NodePath<t.CallExpression>, state) {
				// // `require()` not allowed in esm files
				// // https://nodejs.org/api/esm.html#esm_no_require_exports_or_module_exports
				// if (getFormat(state) === 'mjs' && path.get('callee').isIdentifier({ name: 'require' })) {
				// 	throw new Error(
				// 		`Found a \`require()\` call in non-module file "${paths.basename(
				// 			state.filename,
				// 		)}". Use dynamic \`import()\` instead.`,
				// 	);
				// }

				// `require.resolve()` not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_require_resolve
				if (
					getFormat(state) === 'mjs' &&
					path.get('callee').isMemberExpression() &&
					(path.get('callee.object') as NodePath).isIdentifier({ name: 'require' }) &&
					(path.get('callee.property') as NodePath).isIdentifier({ name: 'resolve' })
				) {
					throw new Error(
						`Found a \`require.resolve()\` call in non-module file "${paths.basename(
							state.filename,
						)}". Use the \`resolve\` npm package instead.`,
					);
				}
			},

			Identifier(path: NodePath<t.Identifier>, state) {
				// __filename -> import.meta.url
				// https://nodejs.org/api/esm.html#esm_no_filename_or_dirname
				if (getFormat(state) === 'mjs' && path.isIdentifier({ name: '__filename' })) {
					path.replaceWith(
						t.memberExpression(
							t.metaProperty(t.identifier('import'), t.identifier('meta')),
							t.identifier('url'),
						),
					);
				}

				// __dirname -> path.dirname(import.meta.url)
				// https://nodejs.org/api/esm.html#esm_no_filename_or_dirname
				if (getFormat(state) === 'mjs' && path.isIdentifier({ name: '__dirname' })) {
					// eslint-disable-next-line @typescript-eslint/no-unsafe-call
					this.pathImport ??= addDefault(path, 'path', { nameHint: '_path' });

					const call = t.callExpression(
						t.memberExpression(
							t.identifier((this.pathImport as { name: string }).name),
							t.identifier('dirname'),
						),
						[
							t.memberExpression(
								t.metaProperty(t.identifier('import'), t.identifier('meta')),
								t.identifier('url'),
							),
						],
					);

					path.replaceWith(call);
				}

				// `NODE_PATH` is not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_node_path
				if (
					getFormat(state) === 'mjs' &&
					path.isIdentifier({ name: 'NODE_PATH' }) &&
					isProcessEnv(path.parentPath)
				) {
					throw new Error(
						'Environment variable `process.env.NODE_PATH` is not available in modules.',
					);
				}
			},

			MemberExpression(path: NodePath<t.MemberExpression>, state) {
				// const isEsm = isEsmFile(state.filename);
				// const file = paths.basename(state.filename);

				// // `exports.<name>` not allowed in esm files
				// // https://nodejs.org/api/esm.html#esm_no_require_exports_or_module_exports
				// if (
				// 	isEsm &&
				// 	getFormat(state) === 'mjs' &&
				// 	path.get('object').isIdentifier({ name: 'exports' }) &&
				// 	path.get('property').isIdentifier()
				// ) {
				// 	console.log(state, this);
				// 	const { name } = path.node.property as t.Identifier;

				// 	throw new Error(
				// 		`Found an \`exports.${name} =\` expression in non-module file "${file}". Use \`export const ${name} =\` instead.`,
				// 	);
				// }

				// // `module.exports` not allowed in esm files
				// // https://nodejs.org/api/esm.html#esm_no_require_exports_or_module_exports
				// if (
				// 	isEsm &&
				// 	getFormat(state) === 'mjs' &&
				// 	path.get('object').isIdentifier({ name: 'module' }) &&
				// 	path.get('property').isIdentifier({ name: 'exports' })
				// ) {
				// 	console.log(state, this);
				// 	throw new Error(
				// 		`Found a \`module.exports =\` expression in non-module file "${file}". Use \`export default\` instead.`,
				// 	);
				// }

				// import.meta.url -> __filename
				// path.dirname(import.meta.url) -> __dirname
				// https://nodejs.org/api/esm.html#esm_no_filename_or_dirname
				if (
					getFormat(state) === 'cjs' &&
					path.get('object').isMetaProperty() &&
					(path.get('object.meta') as NodePath).isIdentifier({ name: 'import' }) &&
					(path.get('object.property') as NodePath).isIdentifier({ name: 'meta' }) &&
					path.get('property').isIdentifier({ name: 'url' })
				) {
					if (isPathDirname(path.parentPath)) {
						path.parentPath.replaceWith(wrapWithFileProto(path.parentPath, '__dirname'));
					} else {
						path.replaceWith(wrapWithFileProto(path, '__filename'));
					}
				}

				// `require.extensions` is not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_require_extensions
				if (
					getFormat(state) === 'mjs' &&
					path.get('object').isIdentifier({ name: 'require' }) &&
					path.get('property').isIdentifier({ name: 'extensions' })
				) {
					throw new Error('API `require.extensions` is not available in modules.');
				}

				// `require.cache` is not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_require_cache
				if (
					getFormat(state) === 'mjs' &&
					path.get('object').isIdentifier({ name: 'require' }) &&
					path.get('property').isIdentifier({ name: 'cache' })
				) {
					throw new Error('API `require.cache` is not available in modules.');
				}

				// `require.resolve` is not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_require_resolve
				if (
					getFormat(state) === 'mjs' &&
					path.get('object').isIdentifier({ name: 'require' }) &&
					path.get('property').isIdentifier({ name: 'resolve' })
				) {
					throw new Error(
						`Found a \`require.resolve()\` call in non-module file "${paths.basename(
							state.filename,
						)}". Use the \`resolve\` npm package instead.`,
					);
				}
			},
		},
	};
}
