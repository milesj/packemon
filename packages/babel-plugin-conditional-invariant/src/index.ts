import { NodePath, types as t } from '@babel/core';

function isWrappedWithConditional(path: NodePath<t.ExpressionStatement>): boolean {
	let currentPath: NodePath<t.Node> | null = path;
	let wrapped = false;

	while (currentPath) {
		// Is there a cleaner way of doing this???
		if (
			currentPath.isIfStatement() &&
			currentPath.get('test').isBinaryExpression({ operator: '!==' }) &&
			(currentPath.get('test.left') as NodePath<t.MemberExpression>).isMemberExpression() &&
			(currentPath.get('test.left.object') as NodePath<t.MemberExpression>).isMemberExpression() &&
			(currentPath.get('test.left.object.object') as NodePath<t.Identifier>).isIdentifier({
				name: 'process',
			}) &&
			(currentPath.get('test.left.object.property') as NodePath<t.Identifier>).isIdentifier({
				name: 'env',
			}) &&
			(currentPath.get('test.left.property') as NodePath<t.Identifier>).isIdentifier({
				name: 'NODE_ENV',
			}) &&
			(currentPath.get('test.right') as NodePath<t.StringLiteral>).isStringLiteral({
				value: 'production',
			})
		) {
			wrapped = true;
			break;
		}

		currentPath = currentPath.parentPath;
	}

	return wrapped;
}

export default function conditionalInvariantPlugin() {
	return {
		visitor: {
			ExpressionStatement: {
				enter(path: NodePath<t.ExpressionStatement>) {
					if (
						!path.get('expression').isCallExpression() ||
						!(path.get('expression.callee') as NodePath<t.Identifier>).isIdentifier({
							name: 'invariant',
						})
					) {
						return;
					}

					// Only transform invariants within the following scenarios:
					// 	- Block statements (if, while, functions, etc)
					// 	- Switch case statements
					//	- Module scope
					if (
						!path.parentPath.isBlockStatement() &&
						!path.parentPath.isSwitchCase() &&
						!path.parentPath.isProgram()
					) {
						return;
					}

					// Dont replace with new syntax if invariant is already wrapped
					// in a NODE_ENV conditional!
					if (isWrappedWithConditional(path)) {
						return;
					}

					path.replaceWith(
						t.ifStatement(
							t.binaryExpression(
								'!==',
								t.memberExpression(
									t.memberExpression(t.identifier('process'), t.identifier('env'), false),
									t.identifier('NODE_ENV'),
									false,
								),
								t.stringLiteral('production'),
							),
							t.blockStatement([path.node]),
						),
					);
				},
			},
		},
	};
}
