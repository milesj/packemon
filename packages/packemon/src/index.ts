/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './Artifact';
export * from './babel/config';
export * from './CodeArtifact';
export * from './commands/Base';
export * from './commands/Build';
export * from './commands/Clean';
export * from './commands/Init';
export * from './commands/Pack';
export * from './commands/Scaffold';
export * from './commands/Validate';
export * from './commands/Watch';
export * from './constants';
export * from './Package';
export * from './PackageValidator';
export * from './Packemon';
export * from './Project';
export * from './rollup/config';
export * from './types';
export * from './TypesArtifact';
