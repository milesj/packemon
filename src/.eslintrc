{
  "rules": {
    "no-console": "off",
    "no-magic-numbers": "off",
    "no-use-before-define": "off",
    "sort-keys": "off",
    "promise/always-return": "off"
  }
}
