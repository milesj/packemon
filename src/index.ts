/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Packemon from './Packemon';
import Package from './Package';
import PackageValidator from './PackageValidator';
import Project from './Project';
import Artifact from './Artifact';
import BundleArtifact from './BundleArtifact';
import TypesArtifact from './TypesArtifact';
import { getBabelInputConfig, getBabelOutputConfig } from './babel/config';

export * from './commands/Build';
export * from './commands/Clean';
export * from './commands/Pack';
export * from './commands/Validate';
export * from './commands/Watch';
export * from './constants';
export * from './types';

export {
  Artifact,
  BundleArtifact,
  TypesArtifact,
  getBabelInputConfig,
  getBabelOutputConfig,
  Package,
  PackageValidator,
  Packemon,
  Project,
};
