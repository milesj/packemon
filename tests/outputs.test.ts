import fs from 'fs';
import { Path } from '@boost/common';
import { getFixturePath } from '@boost/test-utils';
import { BundleArtifact, Package, Project } from '../src';
import { createProjectPackage, createSnapshotSpies } from './helpers';

describe('Outputs', () => {
  describe('artifacts', () => {
    const root = new Path(getFixturePath('project-rollup'));
    const snapshots = createSnapshotSpies(root);

    it('builds all the artifacts with rollup', async () => {
      const pkg = createProjectPackage(root);

      const index = new BundleArtifact(pkg, [
        { format: 'lib', platform: 'node', support: 'stable' },
      ]);
      index.platform = 'node';
      index.support = 'stable';
      index.outputName = 'index';
      index.inputFile = 'src/index.ts';

      pkg.addArtifact(index);

      const client = new BundleArtifact(pkg, [
        { format: 'lib', platform: 'browser', support: 'legacy' },
        { format: 'esm', platform: 'browser', support: 'stable' },
        { format: 'umd', platform: 'browser', support: 'experimental' },
      ]);
      client.platform = 'browser';
      client.outputName = 'client';
      client.inputFile = 'src/client/index.ts';
      client.namespace = 'Packemon';

      pkg.addArtifact(client);

      const server = new BundleArtifact(pkg, [
        { format: 'cjs', platform: 'node', support: 'current' },
      ]);
      server.platform = 'node';
      server.support = 'current';
      server.outputName = 'server';
      server.inputFile = 'src/server/core.ts';

      pkg.addArtifact(server);

      const test = new BundleArtifact(pkg, [
        { format: 'lib', platform: 'native', support: 'experimental' },
      ]);
      test.platform = 'native';
      test.support = 'experimental';
      test.outputName = 'test';
      test.inputFile = 'src/test-utils/base.ts';

      pkg.addArtifact(test);

      try {
        await pkg.build({});
      } catch (error: unknown) {
        console.log(error);
      }

      snapshots().forEach((ss) => {
        expect(ss).toMatchSnapshot();
      });

      expect(index.builds).toMatchSnapshot();
    });
  });

  describe('bundle', () => {
    const root = new Path(getFixturePath('project-bundle'));
    const snapshots = createSnapshotSpies(root);

    it('bundles all files into a single file with rollup', async () => {
      const pkg = createProjectPackage(root);

      const index = new BundleArtifact(pkg, [
        { format: 'lib', platform: 'node', support: 'stable' },
      ]);
      index.bundle = true;
      index.platform = 'node';
      index.support = 'stable';
      index.outputName = 'index';
      index.inputFile = 'src/index.ts';

      pkg.addArtifact(index);

      try {
        await pkg.build({});
      } catch (error: unknown) {
        console.log(error);
      }

      snapshots().forEach((ss) => {
        expect(ss).toMatchSnapshot();
      });
    });
  });

  describe('no bundle', () => {
    const root = new Path(getFixturePath('project-bundle'));
    const snapshots = createSnapshotSpies(root);

    it('creates individual files for every source file', async () => {
      const pkg = createProjectPackage(root);

      const index = new BundleArtifact(pkg, [
        { format: 'lib', platform: 'node', support: 'stable' },
      ]);
      index.bundle = false;
      index.platform = 'node';
      index.support = 'stable';
      index.outputName = 'index';
      index.inputFile = 'src/index.ts';

      pkg.addArtifact(index);

      try {
        // We need to reset to avoid the `entryFileNames` conditional
        process.env.NODE_ENV = 'development';

        await pkg.build({});
      } catch (error: unknown) {
        console.log(error);
      } finally {
        process.env.NODE_ENV = 'test';
      }

      snapshots().forEach((ss) => {
        expect(ss).toMatchSnapshot();
      });
    });
  });
});
