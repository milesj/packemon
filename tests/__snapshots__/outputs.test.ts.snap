// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outputs artifacts builds all the artifacts with rollup 1`] = `
Array [
  "/cjs/server.cjs",
  "// Generated with Packemon: https://packemon.dev
// Platform: node, Support: current, Format: cjs
'use strict';

const https = require('https');

function _interopDefaultLegacy(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    'default': e
  };
}

const https__default = /*#__PURE__*/_interopDefaultLegacy(https);

https__default['default'].createServer((req, res) => {
  res.writeHead(200, {
    'Content-Type': 'text/plain'
  });
  res.write('Hello World!');
  res.end();
}).listen(8080);
//# sourceMappingURL=server.cjs.map
",
]
`;

exports[`Outputs artifacts builds all the artifacts with rollup 2`] = `
Array [
  "/esm/client.js",
  "function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== \\"undefined\\" && o[Symbol.iterator] || o[\\"@@iterator\\"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === \\"number\\") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError(\\"Invalid attempt to iterate non-iterable instance.\\\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.\\"); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === \\"string\\") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === \\"Object\\" && o.constructor) n = o.constructor.name; if (n === \\"Map\\" || n === \\"Set\\") return Array.from(o); if (n === \\"Arguments\\" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

import _regeneratorRuntime from \\"@babel/runtime/regenerator\\";

var createClient = function createClient() {
  try {
    return Promise.resolve(wait()).then(function () {
      var _iterator = _createForOfIteratorHelper(gen()),
          _step;

      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var iterator = _step.value;
          console.log(iterator);
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }

      return new Client();
    });
  } catch (e) {
    return Promise.reject(e);
  }
};

var wait = function wait() {
  return Promise.resolve();
};

var _marked = /*#__PURE__*/_regeneratorRuntime.mark(gen);

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\\"Cannot call a class as a function\\"); } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Generated with Packemon: https://packemon.dev
// Platform: browser, Support: stable, Format: esm
var Client = function Client() {
  _classCallCheck(this, Client);

  _defineProperty(this, \\"prop\\", 123);
};

function gen() {
  return _regeneratorRuntime.wrap(function gen$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
        case \\"end\\":
          return _context.stop();
      }
    }
  }, _marked);
}

export { createClient };
//# sourceMappingURL=client.js.map
",
]
`;

exports[`Outputs artifacts builds all the artifacts with rollup 3`] = `
Array [
  "/lib/client.js",
  "// Generated with Packemon: https://packemon.dev
// Platform: browser, Support: legacy, Format: lib
'use strict';

var _regeneratorRuntime = require(\\"@babel/runtime/regenerator\\");

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== \\"undefined\\" && o[Symbol.iterator] || o[\\"@@iterator\\"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === \\"number\\") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError(\\"Invalid attempt to iterate non-iterable instance.\\\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.\\"); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === \\"string\\") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === \\"Object\\" && o.constructor) n = o.constructor.name; if (n === \\"Map\\" || n === \\"Set\\") return Array.from(o); if (n === \\"Arguments\\" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var createClient = function createClient() {
  try {
    return Promise.resolve(wait()).then(function () {
      var _iterator = _createForOfIteratorHelper(gen()),
          _step;

      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var iterator = _step.value;
          console.log(iterator);
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }

      return new Client();
    });
  } catch (e) {
    return Promise.reject(e);
  }
};

var wait = function wait() {
  return Promise.resolve();
};

var _marked = /*#__PURE__*/_regeneratorRuntime.mark(gen);

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\\"Cannot call a class as a function\\"); } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

Object.defineProperty(exports, '__esModule', {
  value: true
});

var Client = function Client() {
  _classCallCheck(this, Client);

  _defineProperty(this, \\"prop\\", 123);
};

function gen() {
  return _regeneratorRuntime.wrap(function gen$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
        case \\"end\\":
          return _context.stop();
      }
    }
  }, _marked);
}

exports.createClient = createClient;
//# sourceMappingURL=client.js.map
",
]
`;

exports[`Outputs artifacts builds all the artifacts with rollup 4`] = `
Array [
  "/lib/index.js",
  "// Generated with Packemon: https://packemon.dev
// Platform: node, Support: stable, Format: lib
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const EXAMPLE = 'foo';
exports.EXAMPLE = EXAMPLE;
//# sourceMappingURL=index.js.map
",
]
`;

exports[`Outputs artifacts builds all the artifacts with rollup 5`] = `
Array [
  "/lib/test.js",
  "// Generated with Packemon: https://packemon.dev
// Platform: native, Support: experimental, Format: lib
'use strict';

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

Object.defineProperty(exports, '__esModule', {
  value: true
});

class Spy {
  constructor() {
    _defineProperty(this, \\"called\\", false);

    _defineProperty(this, \\"mock\\", () => {});
  }

  method() {}

}

function spy() {
  const inst = new Spy();
  return jest.fn(inst.mock);
}

async function wait() {}

exports.spy = spy;
exports.wait = wait;
//# sourceMappingURL=test.js.map
",
]
`;

exports[`Outputs artifacts builds all the artifacts with rollup 6`] = `
Array [
  "/package.json",
  Object {
    "dependencies": Object {
      "typescript": "*",
    },
    "files": Array [
      "cjs/**/*.{cjs,map}",
      "esm/**/*.{js,map}",
      "lib/**/*.{js,map}",
      "src/**/*.{ts,tsx,json}",
      "umd/**/*.{js,map}",
    ],
    "main": "./lib/index.js",
    "name": "project-rollup",
    "packemon": Object {
      "inputs": Object {
        "client": "src/client.ts",
        "index": "src/index.ts",
        "server": "src/server/core.ts",
        "test": "src/test-utils/base.ts",
      },
    },
  },
]
`;

exports[`Outputs artifacts builds all the artifacts with rollup 7`] = `
Array [
  "/umd/client.js",
  "(function (global, factory) {
  if (typeof define === \\"function\\" && define.amd) {
    define(\\"Packemon\\", [\\"exports\\"], factory);
  } else if (typeof exports !== \\"undefined\\") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.Packemon = mod.exports;
  }
})(typeof globalThis !== \\"undefined\\" ? globalThis : typeof self !== \\"undefined\\" ? self : this, function (_exports) {
  \\"use strict\\";

  Object.defineProperty(_exports, \\"__esModule\\", {
    value: true
  });
  _exports.createClient = createClient;

  function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

  // Generated with Packemon: https://packemon.dev
  // Platform: browser, Support: experimental, Format: umd
  class Client {
    constructor() {
      _defineProperty(this, \\"prop\\", 123);
    }

  }

  function* gen() {}

  async function wait() {}

  async function createClient() {
    await wait();

    for (const iterator of gen()) {
      console.log(iterator);
    }

    return new Client();
  }
});
//# sourceMappingURL=client.js.map
",
]
`;

exports[`Outputs artifacts builds all the artifacts with rollup 8`] = `
Array [
  Object {
    "format": "lib",
    "platform": "node",
    "stats": Object {
      "size": 227,
    },
    "support": "stable",
  },
]
`;

exports[`Outputs bundle bundles all files into a single file with rollup 1`] = `
Array [
  "/lib/index.js",
  "// Generated with Packemon: https://packemon.dev
// Platform: node, Support: stable, Format: lib
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function foo() {}

function bar() {}

exports.bar = bar;
exports.foo = foo;
//# sourceMappingURL=index.js.map
",
]
`;

exports[`Outputs bundle bundles all files into a single file with rollup 2`] = `
Array [
  "/package.json",
  Object {
    "files": Array [
      "lib/**/*.{js,map}",
      "src/**/*.{ts,tsx,json}",
    ],
    "main": "./lib/index.js",
    "name": "project-bundle",
    "packemon": Object {
      "bundle": true,
      "platform": "node",
    },
  },
]
`;

exports[`Outputs no bundle creates individual files for every source file 1`] = `
Array [
  "/lib/index.js",
  "'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var other = require('./other.js');

function bar() {}

exports.foo = other.foo;
exports.bar = bar;
//# sourceMappingURL=index.js.map
",
]
`;

exports[`Outputs no bundle creates individual files for every source file 2`] = `
Array [
  "/lib/not-imported.js",
  "'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function baz() {}

exports.baz = baz;
//# sourceMappingURL=not-imported.js.map
",
]
`;

exports[`Outputs no bundle creates individual files for every source file 3`] = `
Array [
  "/lib/other.js",
  "'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function foo() {}

exports.foo = foo;
//# sourceMappingURL=other.js.map
",
]
`;

exports[`Outputs no bundle creates individual files for every source file 4`] = `
Array [
  "/package.json",
  Object {
    "files": Array [
      "lib/**/*.{js,map}",
      "src/**/*.{ts,tsx,json}",
    ],
    "main": "./lib/index.js",
    "name": "project-bundle",
    "packemon": Object {
      "bundle": true,
      "platform": "node",
    },
  },
]
`;
